runtime/runtime/src/actions.rs

```diff
@@ -624,6 +627,202 @@ pub(crate) fn action_add_key(
+                    account_id: delegate_action.sender_id.clone(),
+                    public_key: delegate_action.public_key.clone(),
+                },
+            )
+            .into());
+            return Ok(());
+        }
+    };
+
+    if delegate_action.nonce <= access_key.nonce {
+        result.result = Err(ActionErrorKind::DelegateActionInvalidNonce {
+            delegate_nonce: delegate_action.nonce,
+            ak_nonce: access_key.nonce,
+        }
+        .into());
+        return Ok(());
+    }
+
+    let upper_bound = apply_state.block_height
+        * near_primitives::account::AccessKey::ACCESS_KEY_NONCE_RANGE_MULTIPLIER;
+    if delegate_action.nonce >= upper_bound {
+        result.result = Err(ActionErrorKind::DelegateActionNonceTooLarge {
+            delegate_nonce: delegate_action.nonce,
+            upper_bound,
+        }
+        .into());
+        return Ok(());
+    }
+
+    access_key.nonce = delegate_action.nonce;
+
+    let actions = delegate_action.get_actions();
+
+    if let AccessKeyPermission::FunctionCall(ref function_call_permission) = access_key.permission {
+        if actions.len() != 1 {
+            result.result = Err(ActionErrorKind::DelegateActionAccessKeyError(
+                InvalidAccessKeyError::RequiresFullAccess,
+            )
+            .into());
+            return Ok(());
+        }
+        if let Some(Action::FunctionCall(ref function_call)) = actions.get(0) {
+            if function_call.deposit > 0 {
+                result.result = Err(ActionErrorKind::DelegateActionAccessKeyError(
+                    InvalidAccessKeyError::DepositWithFunctionCall,
+                )
+                .into());
+            }
+            if delegate_action.receiver_id.as_ref() != function_call_permission.receiver_id {
+                result.result = Err(ActionErrorKind::DelegateActionAccessKeyError(
+                    InvalidAccessKeyError::ReceiverMismatch {
+                        tx_receiver: delegate_action.receiver_id.clone(),
+                        ak_receiver: function_call_permission.receiver_id.clone(),
+                    },
+                )
+                .into());
+                return Ok(());
+            }
+            if !function_call_permission.method_names.is_empty()
+                && function_call_permission
+                    .method_names
+                    .iter()
+                    .all(|method_name| &function_call.method_name != method_name)
+            {
+                result.result = Err(ActionErrorKind::DelegateActionAccessKeyError(
+                    InvalidAccessKeyError::MethodNameMismatch {
+                        method_name: function_call.method_name.clone(),
+                    },
+                )
+                .into());
+                return Ok(());
+            }
+        } else {
+            result.result = Err(ActionErrorKind::DelegateActionAccessKeyError(
```

Comment:
comment under when does this happen (when action is not a function call, right?)